File: server/utils/clientKey.ts
import type { Request } from 'express';

/**
 * 玩家唯一标识：请求头 X-User-Id（浏览器 localStorage 设备 UUID）优先；
 * 无头时回退 IP（同网段会碰撞，仅作兜底）。
 */
export function getClientKey(req: Request): string {
  const userId = req.header('x-user-id')?.trim();
  if (userId) return `uid:${userId.slice(0, 128)}`;

  const forwarded = req.header('x-forwarded-for');
  if (forwarded) {
    const first = forwarded.split(',')[0]?.trim();
    if (first) return `ip:${first}`;
  }

  const realIp = req.header('x-real-ip')?.trim();
  if (realIp) return `ip:${realIp}`;

  const ip = req.ip || req.socket.remoteAddress || 'unknown';
  return `ip:${ip}`;
}

File: server-go/internal/services/client_key.go
package services

import (
	"net/http"
	"strings"
)

// GetClientKey 从请求头派生稳定的玩家标识。
func GetClientKey(r *http.Request) string {
	if userID := strings.TrimSpace(r.Header.Get("X-User-Id")); userID != "" {
		if len(userID) > 128 {
			userID = userID[:128]
		}
		return "uid:" + userID
	}

	if forwarded := r.Header.Get("X-Forwarded-For"); forwarded != "" {
		parts := strings.Split(forwarded, ",")
		if first := strings.TrimSpace(parts[0]); first != "" {
			return "ip:" + first
		}
	}

	if realIP := strings.TrimSpace(r.Header.Get("X-Real-Ip")); realIP != "" {
		return "ip:" + realIP
	}

	ip := r.RemoteAddr
	if host, _, ok := strings.Cut(ip, ":"); ok && strings.Count(ip, ":") == 1 {
		ip = host
	}
	if ip == "" {
		ip = "unknown"
	}
	return "ip:" + ip
}
